test/watch_test.go

```diff
@@ -0,0 +1,200 @@
+				Else(clientv3.OpGet(key)).
+				Commit()
+
+			g.Expect(err).To(BeNil())
+			g.Expect(resp.Succeeded).To(BeTrue())
+		}
+
+		t.Run("Receive", func(t *testing.T) {
+			g := NewWithT(t)
+
+			// receive event
+			g.Eventually(watchCh, testWatchEventPollTimeout).Should(Receive(Satisfy(func(v clientv3.WatchResponse) bool {
+				g.Expect(v.Events).To(HaveLen(1))
+				g.Expect(v.Events[0].Type).To(Equal(clientv3.EventTypePut))
+				g.Expect(v.Events[0].PrevKv).NotTo(BeNil())
+				g.Expect(v.Events[0].PrevKv.Value).To(Equal([]byte(value)))
+				g.Expect(v.Events[0].PrevKv.ModRevision).To(Equal(revAfterCreate))
+
+				g.Expect(v.Events[0].Kv.Key).To(Equal([]byte(key)))
+				g.Expect(v.Events[0].Kv.Value).To(Equal([]byte(updatedValue)))
+				g.Expect(v.Events[0].Kv.Version).To(Equal(int64(0)))
+				g.Expect(v.Events[0].Kv.ModRevision).To(BeNumerically(">", revAfterCreate))
+
+				revAfterUpdate = v.Events[0].Kv.ModRevision
+
+				return true
+			})))
+		})
+
+		t.Run("ReceiveNothingUntilNewActivity", func(t *testing.T) {
+			g := NewWithT(t)
+			g.Consistently(watchCh, testWatchEventIdleTimeout).ShouldNot(Receive())
+		})
+	})
+
+	t.Run("Delete", func(t *testing.T) {
+		g := NewWithT(t)
+
+		// delete key
+		{
+			resp, err := client.Txn(ctx).
+				If(clientv3.Compare(clientv3.ModRevision(key), "=", revAfterUpdate)).
+				Then(clientv3.OpDelete(key)).
+				Else(clientv3.OpGet(key)).
+				Commit()
+
+			g.Expect(err).To(BeNil())
+			g.Expect(resp.Succeeded).To(BeTrue())
+		}
+
+		t.Run("Receive", func(t *testing.T) {
+			g := NewWithT(t)
+
+			// receive event
+			g.Eventually(watchCh, testWatchEventPollTimeout).Should(Receive(Satisfy(func(v clientv3.WatchResponse) bool {
+				g.Expect(v.Events).To(HaveLen(1))
+				g.Expect(v.Events[0].Type).To(Equal(clientv3.EventTypeDelete))
+				g.Expect(v.Events[0].PrevKv).NotTo(BeNil())
+				g.Expect(v.Events[0].PrevKv.Value).To(Equal([]byte(updatedValue)))
+				g.Expect(v.Events[0].PrevKv.ModRevision).To(Equal(revAfterUpdate))
+
+				g.Expect(v.Events[0].Kv).NotTo(BeNil())
+				g.Expect(v.Events[0].Kv.Key).To(Equal([]byte(key)))
+				g.Expect(v.Events[0].Kv.Value).To(Equal([]byte(updatedValue)))
+				g.Expect(v.Events[0].Kv.Version).To(Equal(int64(0)))
+				g.Expect(v.Events[0].Kv.ModRevision).To(BeNumerically(">", revAfterUpdate))
+
+				revAfterDelete = v.Events[0].Kv.ModRevision
+
+				return true
+			})))
+		})
+
+		t.Run("ReceiveNothingUntilNewActivity", func(t *testing.T) {
+			g := NewWithT(t)
+			g.Consistently(watchCh, testWatchEventIdleTimeout).ShouldNot(Receive())
+		})
+	})
+
+	t.Run("StartRevision", func(t *testing.T) {
+		watchAfterDeleteCh := client.Watch(ctx, key, clientv3.WithRev(revAfterUpdate))
+
+		t.Run("Receive", func(t *testing.T) {
+			g := NewWithT(t)
+
+			g.Eventually(watchAfterDeleteCh, testWatchEventPollTimeout).Should(Receive(Satisfy(func(v clientv3.WatchResponse) bool {
+				// receive 2 events
+				g.Expect(v.Events).To(HaveLen(2))
```

Comment:
what are these two events?

Reply:
The lifetime of the `testKey` is this:

| value | revision |
| --| ---|
| _null_ | (before create) |
| testValue | revAfterCreate (e.g. 3) |
| testUpdatedValue | revAfterUpdate (e.g. 4) |
| _null_ | revAfterDelete (e.g. 5) |

when we start a watch from `revAfterUpdate`, we should get the two latest events (with revision number >= the one in the request) -- thus the update and delete events